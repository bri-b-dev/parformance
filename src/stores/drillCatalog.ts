import { defineStore } from 'pinia'
import type { Drill } from '@/types'

export const useDrillCatalogStore = defineStore('drillCatalog', {
  state: () => ({
    drills: [] as Drill[],
    loaded: false,
    error: null as string | null,
  }),
  actions: {
    async load() {
      if (this.loaded && this.drills.length > 0) return
      try {
        // Prefer dynamic import so Vite bundles JSON for production.
        // In the Vitest environment require() the JSON synchronously to avoid async loader timing issues in tests.
        let mod: any
        if (typeof process !== 'undefined' && (process.env.VITEST || process.env.NODE_ENV === 'test')) {
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          mod = require('../data/drills.json')
        } else {
          // @ts-ignore - dynamic import of JSON handled by Vite during build
          mod = await import('../data/drills.json')
        }
        const data = (mod?.default ?? mod) as Drill[]
        if (!Array.isArray(data) || data.length === 0) {
          throw new Error('No drills found in drills.json')
        }
        this.drills = data
        this.error = null
      } catch (e: any) {
        const message = 'Konnte Drills nicht laden. Bitte sp√§ter erneut versuchen.'
        this.error = e?.message || message
        // Friendly toast event for the UI (browser only)
        if (typeof window !== 'undefined' && typeof window.dispatchEvent === 'function') {
          let evt: any
          try {
            // @ts-ignore CustomEvent may be missing in some environments
            evt = new CustomEvent('toast', { detail: { type: 'error', message } })
          } catch {
            evt = { type: 'toast', detail: { type: 'error', message } }
          }
          window.dispatchEvent(evt as any)
        }
      } finally {
        this.loaded = true
      }
    },
  },
})
