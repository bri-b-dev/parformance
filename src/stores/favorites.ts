import { defineStore } from 'pinia'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export const useFavoritesStore = defineStore('favorites', {
  state: () => ({
    favorites: [] as string[],
    loaded: false,
  }),
  getters: {
    isFavorite: (state) => (drillId: string) => state.favorites.includes(drillId),
    list: (state) => state.favorites,
  },
  actions: {
    async load() {
      const stored = persist.get<string[]>('favorites')
      this.favorites = Array.isArray(stored) ? stored : []
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('favorites', this.favorites)
    },
    async add(drillId: string) {
      if (!this.favorites.includes(drillId)) {
        this.favorites.push(drillId)
        await this.persist()
      }
    },
    async remove(drillId: string) {
      const next = this.favorites.filter(id => id !== drillId)
      if (next.length !== this.favorites.length) {
        this.favorites = next
        await this.persist()
      }
    },
    async toggle(drillId: string) {
      if (this.favorites.includes(drillId)) {
        await this.remove(drillId)
      } else {
        await this.add(drillId)
      }
    },
    async clearAll() {
      this.favorites = []
      await this.persist()
    }
  }
})
