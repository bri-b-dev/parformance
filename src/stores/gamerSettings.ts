import { defineStore } from 'pinia'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export type GamerSettings = {
  mode?: string
  handicapOffset?: number
  collapsed?: boolean
}

export const useGamerSettingsStore = defineStore('gamerSettings', {
  state: () => ({
    // map drillId -> settings
    map: {} as Record<string, GamerSettings>,
    loaded: false,
  }),
  getters: {
    forDrill: (state) => (drillId: string) => state.map[drillId] ?? { collapsed: true, mode: undefined, handicapOffset: 0 },
  },
  actions: {
    async load() {
      const stored = persist.get<Record<string, GamerSettings>>('gamerSettings')
      this.map = stored && typeof stored === 'object' ? stored : {}
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('gamerSettings', this.map)
    },
    async setFor(drillId: string, next: GamerSettings) {
      this.map = { ...this.map, [drillId]: { ...(this.map[drillId] ?? {}), ...next } }
      await this.persist()
    },
    async setCollapsed(drillId: string, collapsed: boolean) {
      await this.setFor(drillId, { collapsed })
    },
    async setMode(drillId: string, mode?: string) {
      await this.setFor(drillId, { mode })
    },
    async setHandicapOffset(drillId: string, handicapOffset?: number) {
      await this.setFor(drillId, { handicapOffset })
    }
  }
})
