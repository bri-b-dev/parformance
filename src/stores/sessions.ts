import { defineStore } from 'pinia'
import type { Session } from '@/types'
import { Preferences } from '@capacitor/preferences'

const STORAGE_KEY = 'parformance.sessions.v1'
const isBrowser = typeof window !== 'undefined'

export const useSessionsStore = defineStore('sessions', {
  state: () => ({
    sessions: [] as Session[],
    loaded: false,
  }),
  getters: {
    // Returns all sessions for a given drillId (most recent first)
    listByDrill: (state) => (drillId: string): Session[] =>
      state.sessions
        .filter(s => s.drillId === drillId)
        .sort((a, b) => b.date.localeCompare(a.date)),

    // Returns the latest session for a drillId (by date), or undefined
    latestByDrill(): (drillId: string) => Session | undefined {
      return (drillId: string) => {
        const list = this.listByDrill(drillId)
        return list[0]
      }
    },
  },
  actions: {
    async load() {
      if (!isBrowser) return // skip in non-browser envs (e.g., tests/SSR)
      const stored = await Preferences.get({ key: STORAGE_KEY })
      this.sessions = stored.value ? JSON.parse(stored.value) as Session[] : []
      this.loaded = true
    },
    async persist() {
      if (!isBrowser) return
      await Preferences.set({ key: STORAGE_KEY, value: JSON.stringify(this.sessions) })
    },
    // Create or insert a session. If id is missing, it will be generated.
    async addSession(session: Omit<Session, 'id'> & Partial<Pick<Session, 'id'>>) {
      const id = session.id ?? crypto.randomUUID()
      const toAdd: Session = { ...session, id }
      this.sessions.push(toAdd)
      // Keep sessions roughly ordered by date desc to make getters cheap (not required)
      this.sessions.sort((a, b) => b.date.localeCompare(a.date))
      await this.persist()
      return toAdd
    },
    async remove(id: string) {
      this.sessions = this.sessions.filter(s => s.id !== id)
      await this.persist()
    },
    async clearAll() {
      this.sessions = []
      await this.persist()
    }
  }
})
