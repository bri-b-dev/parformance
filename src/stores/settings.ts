import { defineStore } from 'pinia'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export type Handedness = 'right' | 'left'

export interface SettingsState {
  hcp: number | null
  handedness: Handedness
  loaded: boolean
}

export const useSettingsStore = defineStore('settings', {
  state: (): SettingsState => ({
    hcp: null,
    handedness: 'right',
    loaded: false,
  }),
  getters: {
    // direct getters if needed in components
    getHcp: (state) => state.hcp,
    getHandedness: (state) => state.handedness,
  },
  actions: {
    async load() {
      const data = persist.get<Partial<SettingsState>>('settings')
      if (data) {
        if (typeof data.hcp === 'number' || data.hcp === null) this.hcp = data.hcp ?? null
        if (data.handedness === 'right' || data.handedness === 'left') this.handedness = data.handedness
      }
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('settings', { hcp: this.hcp, handedness: this.handedness })
    },
    async setHcp(hcp: number | null) {
      this.hcp = hcp
      // Emit a browser event so components can re-render HCP highlights
      if (typeof window !== 'undefined' && typeof window.dispatchEvent === 'function') {
        let evt: any
        try {
          // Prefer CustomEvent when available (browser)
          // @ts-ignore - CustomEvent may not exist in Node
          evt = new CustomEvent('hcp-changed', { detail: hcp })
        } catch {
          // Fallback for non-browser test environments
          evt = { type: 'hcp-changed', detail: hcp }
        }
        window.dispatchEvent(evt as any)
      }
      await this.persist()
    },
    async setHandedness(value: Handedness) {
      this.handedness = value
      await this.persist()
    },
    async clearAll() {
      this.hcp = null
      this.handedness = 'right'
      await this.persist()
    },
  },
})
