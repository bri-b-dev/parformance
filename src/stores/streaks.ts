import { defineStore } from 'pinia'
import { useSessionsStore } from './sessions'
import type { Session } from '@/types'
import { computeStreaks } from '@/utils/streaks'

export const useStreaksStore = defineStore('streaks', {
  getters: {
    stats(): { current: number; best: number } {
      const sessionsStore = useSessionsStore()
      const sessions = sessionsStore.sessions as Session[]
      return computeStreaks(
        sessions.map(s => ({ date: s.date }))
      )
    },
    current(): number {
      return this.stats.current
    },
    best(): number {
      return this.stats.best
    }
  },
})

export default { useStreaksStore }
