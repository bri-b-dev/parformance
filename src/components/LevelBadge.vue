<template>
  <span
    class="chip"
    :title="tooltip"
    :aria-label="ariaLabel"
    data-testid="level-badge"
  >
    <template v-if="level != null">
      Level {{ level }}
    </template>
    <template v-else>
      Level —
    </template>
  </span>
</template>

<script setup lang="ts">
import { computed, onMounted } from 'vue'
import type { Drill } from '@/types'
import { useSessionsStore } from '@/stores/sessions'
import { useSettingsStore } from '@/stores/settings'
import { computeLevelForDrill } from '@/hcp/level'
import { findBucketKey, formatRangeLabel } from '@/hcp/buckets'

const props = defineProps<{ drill: Drill }>()

const sessions = useSessionsStore()
const settings = useSettingsStore()

onMounted(async () => {
  if (!sessions.loaded) await sessions.load()
  if (!settings.loaded) await settings.load()
})

// Latest session for this drill
const latest = computed(() => sessions.latestByDrill(props.drill.id))

// Display level: prefer stored levelReached; otherwise compute from current hcp + result
const level = computed<1|2|3|null>(() => {
  const s = latest.value
  if (!s) return null
  let lvl = typeof s.levelReached === 'number' ? s.levelReached : computeLevelForDrill(props.drill, settings.hcp, Number(s.result?.value))
  if (!Number.isFinite(lvl)) return null
  // Only render 1..3; 0 maps to neutral
  if (lvl >= 1 && lvl <= 3) return lvl as 1|2|3
  return null
})

const tooltip = computed(() => {
  const targets = props.drill.metric?.hcpTargets || {}
  const key = typeof settings.hcp === 'number' ? findBucketKey(settings.hcp, targets) : null
  if (!key) {
    return 'Keine HCP-Zielwerte vorhanden'
  }
  const arr = targets[key] || []
  const [L1, L2, L3] = [arr[0], arr[1], arr[2]]
  const label = formatRangeLabel(key)
  const parts: string[] = []
  if (Number.isFinite(L1)) parts.push(`L1 ${L1}`)
  if (Number.isFinite(L2)) parts.push(`L2 ${L2}`)
  if (Number.isFinite(L3)) parts.push(`L3 ${L3}`)
  return parts.length ? `HCP ${label}: ${parts.join(', ')}` : `HCP ${label}: —`
})

const ariaLabel = computed(() => level.value ? `Level erreicht: ${level.value}` : 'Kein Level ermittelt')
</script>
