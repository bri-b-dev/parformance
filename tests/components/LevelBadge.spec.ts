import { describe, it, expect, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import LevelBadge from '../../src/components/LevelBadge.vue'
import type { Drill, Session } from '../../src/types'
import { useSessionsStore } from '../../src/stores/sessions'
import { useSettingsStore } from '../../src/stores/settings'

const drill: Drill = {
  id: 'chip_carry_zone',
  title: 'Carry Zone',
  category: 'Chipping Green',
  equipment: {},
  setup: { schema: 'x' },
  duration: {},
  instructions: { training: 'x' },
  metric: { type: 'streak', unit: 'Zonen in Serie', hcpTargets: { '54-27': [2,3,4], '26-12': [5,7,9], '11-0': [11,13,15] } },
}

function seedStores({ sessions = [], hcp = 18 }: { sessions?: Session[]; hcp?: number } = {}) {
  const sess = useSessionsStore()
  const sets = useSettingsStore()
  // Seed state and mark as loaded so component's onMounted won't call load()
  sess.sessions = sessions
  ;(sess as any).loaded = true
  sets.hcp = hcp
  ;(sets as any).loaded = true
}

describe('LevelBadge', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
  })

  it('renders neutral when there are no sessions and shows thresholds in tooltip (snapshot)', async () => {
    seedStores({ sessions: [], hcp: 18 })
    const wrapper = mount(LevelBadge, { props: { drill } })
    const badge = wrapper.get('[data-testid="level-badge"]')
    expect(badge.text()).toContain('Level —')
    // Tooltip should mention the HCP bucket and thresholds for 26–12
    const title = badge.attributes('title') || ''
    expect(title).toMatch(/HCP 26–12/)
    expect(title).toMatch(/L1 5/)
    expect(title).toMatch(/L2 7/)
    expect(title).toMatch(/L3 9/)

    // Snapshot of neutral rendering
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders latest level (1–3) when a recent session exists', async () => {
    const sessions: Session[] = [
      { id: 's1', drillId: drill.id, date: '2025-01-01T10:00:00Z', hcp: 18, result: { value: 4, unit: drill.metric.unit }, levelReached: 0 },
      { id: 's2', drillId: drill.id, date: '2025-02-01T10:00:00Z', hcp: 18, result: { value: 8, unit: drill.metric.unit }, levelReached: 2 },
    ]
    seedStores({ sessions, hcp: 18 })

    const wrapper = mount(LevelBadge, { props: { drill } })
    const badge = wrapper.get('[data-testid="level-badge"]')
    expect(badge.text()).toContain('Level 2')
  })
})
