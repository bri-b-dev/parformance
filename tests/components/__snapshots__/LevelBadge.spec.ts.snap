// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LevelBadge > renders neutral when there are no sessions and shows thresholds in tooltip (snapshot) 1`] = `"<span class="chip" title="HCP 26–12: L1 5, L2 7, L3 9" aria-label="Kein Level ermittelt" data-testid="level-badge"> Level — </span>"`;
